### PR TITLE
Moving mutated php code to it's own file

### PR DESCRIPTION
**Pull request by deleugpn**
Hey guys, first I'd like to apologize for my PR #500 which is quite late already. I'm trying to better organize my time to finish that one.
Today I got a spare minute and tried to fix the conflict that it has, but it was quite hard for me. Not only that, but also when I was actually working on adding the new mutations there I noticed the difficulty of adding new mutated code into the test files.

This PR is a Proof Of Concept proposal to use the `getFileContent` method not only for the original code, but also for the mutated one. The mutated code would be suffixed by `mx` where `x` is the number of the mutation (e.g. `pv-final-m0.php`, `pv-final-m1.php`, etc...). The reason for the suffix is that with the `yield` system we might generate more than 1 mutation per original code.

If this gets a positive feedback, I'll finish moving the entire file to this system and then rework PR #500 which would be a lot easier.

### CODE REVIEW

**Comment by borNfreee** [comment]
Hi, I'm not sure I understood what problem this update is trying to solve. So as I see from the description, you are experiencing 2 issues:

1. It's hard to merge conflicts when the mutated source code is inside the test file, but not in a separate one.
2. It's hard to add new test cases

Questions:

1. How does moving the code in a separate file simplify merging conflicts? 
2. Why do you think that adding a new file is easier than adding some lines of code to the test method?

As for me, it's very convenient to see the output of mutation just looking at the test case. No need to open a file. DX is better. 

Would you mind give some more details?

**Comment by deleugpn** [comment]
1) By having the mutated content in separate files, a merge conflict only has to deal with a small simple php file. Right now, I have a merge conflict that was done in a file that only had 1 mutation and I changed most of it to 2 mutations. So I have to deal with a lot of array commas and it's incredibly hard to find the right place for the file content, the inner array separator (mutations) and the outer array separator (mutated, mutators). By moving things to a separate file, any conflict will be, at worst, a one-line conflict change similar to `$this->getFileContents($param),` and if there is a conflict inside the mutated file, there's no mix between PHP File vs Inline PHP Content.

2) Pretty much the same thing listed on 1). This is my second time adding multiple mutations to a test code and it usually goes like this: 

### Original Content
```
    public function provideMutationCases(): \Generator
    {
        yield 'It removes casting to bool with "bool"' => [
            <<<'PHP'
<?php

(bool) 1;
PHP
            ,
            <<<'PHP'
<?php

1;
PHP
            ,
        ];

        yield 'It removes casting to bool with "boolean"' => [
            <<<'PHP'
<?php

(boolean) 1;
PHP
            ,
            <<<'PHP'
<?php

1;
PHP
            ,
        ];
    }
```

1- Find the test divisors

``` 
        yield 'It removes casting to bool with "bool"' => [
...
        yield 'It removes casting to bool with "boolean"' => [
...
```

2- Wrap the mutation in another array.

```
    public function provideMutationCases(): \Generator
    {
        yield 'It removes casting to bool with "bool"' => [
            <<<'PHP'
<?php

(bool) 1;
PHP
            , [    // <--------------
            <<<'PHP'
<?php

1;
PHP
            ,
            ],    // <--------------
        ];

        yield 'It removes casting to bool with "boolean"' => [
            <<<'PHP'
<?php

(boolean) 1;
PHP
            ,
            [   // <--------------
            <<<'PHP'
<?php

1;
PHP
            ,
            ],    // <--------------
        ];
    }
}
```

3- Add another mutation into the inner array

```
    public function provideMutationCases(): \Generator
    {
        yield 'It removes casting to bool with "bool"' => [
            <<<'PHP'
<?php

(bool) 1;
PHP
            , [
            <<<'PHP'
<?php

1;
PHP
            ,
                <<<'PHP'            // <------------
<?php

(string) (1);
PHP                                    // <------------
                ,
            ],
        ];

        yield 'It removes casting to bool with "boolean"' => [
            <<<'PHP'
<?php

(boolean) 1;
PHP
            ,
            [
            <<<'PHP'
<?php

1;
PHP
            , 
            <<<'PHP'           // <------------
<?php                           

(string) 1;
PHP
                ,                     // <------------
            ],
        ];
    }
}
```

And after this, if there is a merge conflict, it's likely that the previous commit was done before we had inner arrays, so it's especially harder to find where to place the new test scenario in the array of arrays of arrays.

My proposal brings something like this: 

```
    public function provideMutationCases(): \Generator
    {
        yield 'It removes casting to bool with "bool"' => [
            $this->fileGetContents('bool-casting'), [
                $this->fileGetContents('bool-casting-m0'),
                $this->fileGetContents('bool-casting-m1'),
            ],
        ];

        yield 'It removes casting to bool with "boolean"' => [
            $this->fileGetContents('boolean-casting'), [
                $this->fileGetContents('boolean-casting-m0'),
                $this->fileGetContents('boolean-casting-m1'),
            ],
        ];
    }
```

It's extremely easy to identify the outer array and the inner array while removing any possible cognitive mix between the PHP File and the Inline PHP Content.

File: tests/Fixtures/Autoloaded/PublicVisibility/pv-final-m0.php [hunk 1]
```diff
@@ -0,0 +1,12 @@
...
+namespace PublicVisibilityFinal;
+
+class Test
+{
+    protected final function foo(int $param, $test = 1) : bool
+    {
+        echo 1;
+        return false;
+    }
+}
```
Comment (BackEndTea) [inline tests/Fixtures/Autoloaded/PublicVisibility/pv-final-m0.php]: These, and a lot of other files are missing a new line at the end of the file.

**Comment by deleugpn** [comment]
This is ready for review :)

**Comment by localheinz** [comment]
In order to improve navigability for a developer using an IDE, what do you think about making the following adjustments, allowing to reference full paths instead of using `sprintf()` to build full file paths?

```diff
diff --git a/tests/Mutator/FunctionSignature/PublicVisibilityTest.php b/tests/Mutator/FunctionSignature/PublicVisibilityTest.php
index a4a74cd..5a871c3 100644
--- a/tests/Mutator/FunctionSignature/PublicVisibilityTest.php
+++ b/tests/Mutator/FunctionSignature/PublicVisibilityTest.php
@@ -19,9 +19,9 @@ final class PublicVisibilityTest extends AbstractMutatorTestCase
     /**
      * @dataProvider blacklistedProvider
      */
-    public function test_it_does_not_modify_blacklisted_functions(string $functionName, string $args = '', string $modifier = ''): void
+    public function test_it_does_not_modify_blacklisted_functions(string $file): void
     {
-        $code = $this->getFileContent("pv-{$functionName}.php");
+        $code = file_get_contents($file);

         $this->doTest($code);
     }
@@ -29,16 +29,16 @@ final class PublicVisibilityTest extends AbstractMutatorTestCase
     public function blacklistedProvider(): array
     {
         return [
-            ['__construct'],
-            ['__invoke'],
-            ['__call', '$n, $v'],
-            ['__callStatic', '$n, $v', 'static '],
-            ['__get', '$n'],
-            ['__set', '$n, $v'],
-            ['__isset', '$n'],
-            ['__unset', '$n'],
-            ['__toString'],
-            ['__debugInfo'],
+            '__call' => [__DIR__ . '/../../Fixtures/Autoloaded/PublicVisibility/pv-__call.php'],
+            '__callStatic' => [__DIR__ . '/../../Fixtures/Autoloaded/PublicVisibility/pv-__callStatic.php'],
+            '__construct' => [__DIR__ . '/../../Fixtures/Autoloaded/PublicVisibility/pv-__construct.php'],
+            '__debugInfo' => [__DIR__ . '/../../Fixtures/Autoloaded/PublicVisibility/pv-__debugInfo.php'],
+            '__get' => [__DIR__ . '/../../Fixtures/Autoloaded/PublicVisibility/pv-__get.php'],
+            '__invoke' => [__DIR__ . '/../../Fixtures/Autoloaded/PublicVisibility/pv-__invoke.php'],
+            '__isset' => [__DIR__ . '/../../Fixtures/Autoloaded/PublicVisibility/pv-__isset.php'],
+            '__set' => [__DIR__ . '/../../Fixtures/Autoloaded/PublicVisibility/pv-__set.php'],
+            '__toString' => [__DIR__ . '/../../Fixtures/Autoloaded/PublicVisibility/pv-__toString.php'],
+            '__unset' => [__DIR__ . '/../../Fixtures/Autoloaded/PublicVisibility/pv-__unset.php'],
         ];
     }

@@ -53,76 +53,76 @@ final class PublicVisibilityTest extends AbstractMutatorTestCase
     public function provideMutationCases(): \Generator
     {
         yield 'It mutates public to protected' => [
-            $this->getFileContent('pv-one-class.php'),
+            file_get_contents(__DIR__ . '/../../Fixtures/Autoloaded/PublicVisibility/pv-one-class.php'),
             [
-                $this->getFileContent('pv-one-class-m0.php'),
+                file_get_contents(__DIR__ . '/../../Fixtures/Autoloaded/PublicVisibility/pv-one-class-m0.php'),
             ],
         ];

         yield 'It does not mutate final flag' => [
-            $this->getFileContent('pv-final.php'),
+            file_get_contents(__DIR__ . '/../../Fixtures/Autoloaded/PublicVisibility/pv-final.php'),
             [
-                $this->getFileContent('pv-final-m0.php'),
+                file_get_contents(__DIR__ . '/../../Fixtures/Autoloaded/PublicVisibility/pv-final-m0.php'),
             ],
         ];

         yield 'It mutates non abstract public to protected in an abstract class' => [
-            $this->getFileContent('pv-non-abstract-in-abstract-class.php'),
+            file_get_contents(__DIR__ . '/../../Fixtures/Autoloaded/PublicVisibility/pv-non-abstract-in-abstract-class.php'),
             [
-                $this->getFileContent('pv-non-abstract-in-abstract-class-m0.php'),
+                file_get_contents(__DIR__ . '/../../Fixtures/Autoloaded/PublicVisibility/pv-non-abstract-in-abstract-class-m0.php'),
             ],
         ];

         yield 'It does not mutate static flag' => [
-            $this->getFileContent('pv-static.php'),
+            file_get_contents(__DIR__ . '/../../Fixtures/Autoloaded/PublicVisibility/pv-static.php'),
             [
-                $this->getFileContent('pv-static-m0.php'),
+                file_get_contents(__DIR__ . '/../../Fixtures/Autoloaded/PublicVisibility/pv-static-m0.php'),
             ],
         ];

         yield 'It replaces visibility if not set' => [
-            $this->getFileContent('pv-not-set.php'),
+            file_get_contents(__DIR__ . '/../../Fixtures/Autoloaded/PublicVisibility/pv-not-set.php'),
             [
-                $this->getFileContent('pv-not-set-m0.php'),
+                file_get_contents(__DIR__ . '/../../Fixtures/Autoloaded/PublicVisibility/pv-not-set-m0.php'),
             ],
         ];

         yield 'It does not mutate an interface' => [
-            $this->getFileContent('pv-interface.php'),
+            file_get_contents(__DIR__ . '/../../Fixtures/Autoloaded/PublicVisibility/pv-interface.php'),
         ];

         yield 'It does not mutate an abstract function' => [
-            $this->getFileContent('pv-abstract.php'),
+            file_get_contents(__DIR__ . '/../../Fixtures/Autoloaded/PublicVisibility/pv-abstract.php'),
         ];

         yield 'It does not mutate if interface has same public method' => [
-            $this->getFileContent('pv-same-method-interface.php'),
+            file_get_contents(__DIR__ . '/../../Fixtures/Autoloaded/PublicVisibility/pv-same-method-interface.php'),
         ];

         yield 'It does not mutate if any of interfaces has same public method' => [
-            $this->getFileContent('pv-same-method-any-interface.php'),
+            file_get_contents(__DIR__ . '/../../Fixtures/Autoloaded/PublicVisibility/pv-same-method-any-interface.php'),
         ];

         yield 'It does not mutate if parent abstract has same public method' => [
-            $this->getFileContent('pv-same-method-abstract.php'),
+            file_get_contents(__DIR__ . '/../../Fixtures/Autoloaded/PublicVisibility/pv-same-method-abstract.php'),
         ];

         yield 'It does not mutate if parent class has same public method' => [
-            $this->getFileContent('pv-same-method-parent.php'),
+            file_get_contents(__DIR__ . '/../../Fixtures/Autoloaded/PublicVisibility/pv-same-method-parent.php'),
             [
-                $this->getFileContent('pv-same-method-parent-m0.php'),
+                file_get_contents(__DIR__ . '/../../Fixtures/Autoloaded/PublicVisibility/pv-same-method-parent-m0.php'),
             ],
         ];

         yield 'it does not mutate if grandparent class has same public method' => [
-            $this->getFileContent('pv-same-method-grandparent.php'), [
-                $this->getFileContent('pv-same-method-grandparent-m0.php'),
+            file_get_contents(__DIR__ . '/../../Fixtures/Autoloaded/PublicVisibility/pv-same-method-grandparent.php'), [
+                file_get_contents(__DIR__ . '/../../Fixtures/Autoloaded/PublicVisibility/pv-same-method-grandparent-m0.php'),
             ],
         ];

         yield 'it does mutate non-inherited methods' => [
-            $this->getFileContent('pv-non-same-method-parent.php'), [
-                $this->getFileContent('pv-non-same-method-parent-m0.php'),
+            file_get_contents(__DIR__ . '/../../Fixtures/Autoloaded/PublicVisibility/pv-non-same-method-parent.php'), [
+                file_get_contents(__DIR__ . '/../../Fixtures/Autoloaded/PublicVisibility/pv-non-same-method-parent-m0.php'),
             ],
         ];

@@ -142,9 +142,4 @@ function something()
 PHP
         ];
     }
-
-    private function getFileContent(string $file): string
-    {
-        return file_get_contents(sprintf(__DIR__ . '/../../Fixtures/Autoloaded/PublicVisibility/%s', $file));
-    }
 }
```

**Comment by deleugpn** [comment]
I had totally forgot about this PR. It's very likely that this branch is incredibly out of date. I remember spending hours hunting down the broken test and couldn't figure out why it was broken. I'll be closing this for now and hope to be able to come back to it some other time.
Unfortunately I haven't been able to land infection on my company's pipeline yet.

**Comment by borNfreee** [comment]
> Unfortunately I haven't been able to land infection on my company's pipeline yet.

What is the main issue? If it takes too much time, you can run Infection just for the changed files https://infection.github.io/guide/how-to.html#How-to-run-Infection-only-for-changed-files

**Pull request closed without merging**